wcfsetup/install/files/lib/system/flood/FloodControl.class.php

```diff
@@ -0,0 +1,241 @@
+	 */
+	private function countContentByIdentifier(string $objectType, string $identifier, \DateInterval $interval, int $time): array {
+		$sql = "SELECT  COUNT(*) AS count, MIN(time) AS earliestTime
+			FROM    wcf" . WCF_N . "_flood_control
+			WHERE   objectTypeID = ?
+			        AND identifier = ?
+			        AND (time > ? AND time <= ?)";
+		$statement = WCF::getDB()->prepareStatement($sql);
+		$statement->execute([
+			$this->getObjectTypeID($objectType),
+			$identifier,
+			DateUtil::getDateTimeByTimestamp($time)->sub($interval)->getTimestamp(),
+			$time,
+		]);
+		
+		return $statement->fetchSingleRow();
+	}
+	
+	/**
+	 * Returns the number of contents the active user created of the given type within a certain
+	 * period of time `($time-$interval, $time]` and the earliest time within the period content
+	 * was created.
+	 */
+	public function countContent(string $objectType, \DateInterval $interval, int $time = TIME_NOW): array {
+		if (WCF::getUser()->userID) {
+			return $this->countUserContent($objectType, WCF::getUser()->userID, $interval, $time);
+		}
+		else {
+			return $this->countGuestContent($objectType, UserUtil::getIpAddress(), $interval, $time);
+		}
+	}
+	
+	/**
+	 * Returns the number of contents a guest created of the given type within a certain period
+	 * of time `($time-$interval, $time]` and the earliest time within the period content was
+	 * created.
+	 */
+	public function countGuestContent(string $objectType, string $ipAddress, \DateInterval $interval, int $time = TIME_NOW): array {
+		return $this->countContentByIdentifier(
+			$objectType,
+			$this->getUserIdentifier($objectType, $ipAddress),
+			$interval,
+			$time
+		);
+	}
+	
+	/**
+	 * Returns the number of contents a user created of the given type within a certain period
+	 * of time `[$time-$interval, $time]` and the earliest time within the period content was
+	 * created.
+	 */
+	public function countUserContent(string $objectType, int $userID, \DateInterval $interval, int $time = TIME_NOW): array {
+		return $this->countContentByIdentifier(
+			$objectType,
+			$this->getUserIdentifier($objectType, $userID),
+			$interval,
+			$time
+		);
+	}
+	
+	/**
+	 * Returns the identifier used for a guest with the given ip address for content of the
+	 * given object type.
+	 */
+	private function getGuestIdentifier(string $objectType, string $ipAddress): string {
+		return \hash_hmac(
+			'md5',
+			'guest:' . $ipAddress,
+			self::IDENTIFIER_PREFIX . $objectType,
+			true
+		);
+	}
+	
+	/**
+	 * Returns the last time a guest/user created content of the given type or `null` if they
+	 * have not created such content.
+	 */
+	private function getLastTimeByIdentifier(string $objectType, string $identifier): ?int {
+		$sql = "SELECT          time
+			FROM            wcf" . WCF_N . "_flood_control
+			WHERE           objectTypeID = ?
+			                AND identifier = ?
+			ORDER BY        time DESC";
+		$statement = WCF::getDB()->prepareStatement($sql, 1);
```

Comment:
Check the query plan, but I expect the suggestion to look better or same with it being more readable.

```suggestion
		$sql = "SELECT          MAX(time)
			FROM            wcf" . WCF_N . "_flood_control
			WHERE           objectTypeID = ?
			                AND identifier = ?";
		$statement = WCF::getDB()->prepareStatement($sql);
```